lib/quantum/execution_broadcaster.ex

```diff
@@ -0,0 +1,131 @@
+defmodule Quantum.ExecutionBroadcaster do
+  use GenStage
+
+  alias Quantum.Job
+  alias Crontab.{Scheduler, CronExpression}
+
+  @date_library Application.get_env(:quantum, :date_library, Quantum.DateLibrary.Timex)
+
+  def start_link(opts) do
+    case GenStage.start_link(__MODULE__, opts, [name: Keyword.fetch!(opts, :execution_broadcaster)]) do
+      {:ok, pid} ->
+        {:ok, pid}
+      {:error, {:already_started, pid}} ->
+        Process.link(pid)
+        {:ok, pid}
+    end
+  end
+
+  def init(opts) do
+    state = %{jobs: [], time: NaiveDateTime.utc_now, timer: nil}
+    {:producer_consumer, state, subscribe_to: [Keyword.fetch!(opts, :job_broadcaster)]}
+  end
+
+  def handle_events(events, _, state) do
+    reboot_add_events = events
+    |> Enum.filter(&add_reboot_event?/1)
+    |> Enum.map(fn {:add, job} -> {:execute, job} end)
+
+    state = events
+    |> Enum.reject(&add_reboot_event?/1)
+    |> Enum.reduce(state, &handle_event/2)
+    |> sort_state
+    |> reset_timer
+
+    {:noreply, reboot_add_events, state}
+  end
+
+  def handle_info(:execute, %{jobs: [{time_to_execute, jobs_to_execute} | tail]} = state) do
+    state = state
+    |> Map.put(:timer, nil)
+    |> Map.put(:jobs, tail)
+    |> Map.put(:time, NaiveDateTime.add(time_to_execute, 1, :second))
+
+    state = jobs_to_execute
+    |> Enum.reduce(state, &add_job_to_state/2)
+    |> sort_state
+    |> reset_timer
+
+    {:noreply, Enum.map(jobs_to_execute, fn job -> {:execute, job} end), state}
+  end
+
+
+  defp handle_event({:add, job}, state) do
+    add_job_to_state(job, state)
+  end
+
+  defp add_job_to_state(%Job{schedule: schedule, timezone: timezone} = job, %{time: time} = state) do
+    case Scheduler.get_next_run_date(schedule, time) do
+      {:ok, date} ->
+        add_to_state(state, to_utc(date, timezone), job)
+      _ ->
+        state
+    end
+  end
+
+  defp to_utc(date, :utc), do: date
+  defp to_utc(date, tz) when is_binary(tz), do: @date_library.tz_to_utc(date, tz)
+
+  defp sort_state(%{jobs: jobs} = state) do
+    %{state | jobs: Enum.sort(jobs, fn {date_1, _}, {date_2, _} ->
+      NaiveDateTime.to_erl(date_1) < NaiveDateTime.to_erl(date_2)
+    end)}
+  end
+
+  defp add_to_state(%{jobs: jobs} = state, date, job) do
+    %{state | jobs: case Enum.find_index(jobs, fn {run_date, _} -> run_date == date end) do
```

Comment:
`Enum.find_index` and `List.update_at` are both `O(n)`, consider using a `Map`, since you are doing equality check anyway, naturally it will be a map of `run_date`'s to `jobs`. This will make `add_to_state` `O(1)`